include/mbgl/util/padding.hpp

```diff
@@ -0,0 +1,80 @@
+#pragma once
+
+#include <mbgl/util/feature.hpp>
+#include <span>
+
+namespace mbgl {
+
+// A set of four numbers representing padding around a box.
+//
+//
+struct Padding {
+    // Empty/zero padding on all sides.
+    Padding() = default;
+
+    // Same padding on all sides.
+    Padding(float value)
+        : top(value),
+          right(value),
+          bottom(value),
+          left(value) {}
+
+    Padding(float t_, float r_, float b_, float l_)
+        : top(t_),
+          right(r_),
+          bottom(b_),
+          left(l_) {}
+
+    // Padding values are in CSS order: top, right, bottom, left.
+    Padding(const std::span<const float>& values) {
+        switch (values.size()) {
+            case 1:
+                top = right = bottom = left = values[0];
+                break;
+            case 2:
+                top = bottom = values[0];
+                right = left = values[1];
+                break;
+            case 3:
+                top = values[0];
+                left = right = values[1];
+                bottom = values[2];
+                break;
+            case 4:
+                top = values[0];
+                right = values[1];
+                bottom = values[2];
+                left = values[3];
+                break;
+            default:
+                throw std::invalid_argument("Padding must have between 1 and 4 elements");
+        }
+    }
+
+    float top = 0;
```

Comment:
We should make this immutable probably.

Reply:
This breaks constructor from std::span. I don't want to copy this logic in several places where we call it. I

I can make it immutable and add factory method instead of ctor, but it feels like an overkill for such a simple struct with 4 fields. I'd prefer to keep it mutable. Please ping me if you if you feel strongly about this and want me to change this.

Reply:
I tried this:

```cpp
struct Padding {
    // Empty/zero padding on all sides.
    // Padding() = default;

    // Same padding on all sides.
    Padding(float value)
        : top(value),
          right(value),
          bottom(value),
          left(value) {}

    Padding(float t_, float r_, float b_, float l_)
        : top(t_),
          right(r_),
          bottom(b_),
          left(l_) {}

    /// Padding values are in CSS order: top, right, bottom, left.
    Padding(const std::span<const float>& values)
        : Padding(Padding::paddingTupleFromSpan(values)) {}

    const float top = 0;
    const float right = 0;
    const float bottom = 0;
    const float left = 0;

    explicit operator bool() const { return top != 0 || right != 0 || bottom != 0 || left != 0; }

    bool operator==(const Padding&) const = default;

    std::array<float, 4> toArray() const;

    // Used by ValueFactory<Padding>::make()
    mbgl::Value serialize() const;

private:
    // Helper function to compute padding values from a span
    static std::tuple<float, float, float, float> paddingTupleFromSpan(const std::span<const float>& values) {
        switch (values.size()) {
            case 1:
                return {values[0], values[0], values[0], values[0]};
            case 2:
                return {values[0], values[1], values[0], values[1]};
            case 3:
                return {values[0], values[1], values[2], values[1]};
            case 4:
                return {values[0], values[1], values[2], values[3]};
            default:
                throw std::invalid_argument("Padding must have between 1 and 4 elements");
        }
    }

    // Private constructor used by the std::span constructor
    Padding(const std::tuple<float, float, float, float>& padding)
        : top(std::get<0>(padding)),
          right(std::get<1>(padding)),
          bottom(std::get<2>(padding)),
          left(std::get<3>(padding)) {}
};
```

But I am getting some weird compile errors.

Reply:
I was mostly averse to complicating this simple struct, but I love immutable things as the next guy, so I'll look into using this approach! :)

Reply:
Bart, the tuple approach works wonderfully. I don't think I've changed anything, except moving impl to cpp. 

But... immutable Padding causes more changes downstream. There are several places where code either uses assignment operator directly or via autogenerated operators in classes which include Padding as a member (e.g. PositionedIcon). Immutability breaks those.

So far I've updated 6 places, mostly by either adding std::optional wrapper or by replacing `x = foo;` with `x.emplace(foo)` if the object was already an optional.  

Then I got to expressions code and found out that generic code around `defaultValue` expects underlying types to be mutable too. It is probably fixable too, but at this point I doubt that the whole caboodle is worth it and wanted to double-check with you. 

Most (all?) other simple structs I've checked are declared mutable and immutability is added by slapping const on them as necessary. Maybe we should follow the same idea with Padding?

Reply:
Yes exactly I also went down that rabbit hole for a bit. 🙈 

Let's keep it the way it is.